analyzers/src/SonarAnalyzer.Common/Rules/MethodOverloadsShouldBeGroupedBase.cs

```diff
@@ -18,49 +18,57 @@
         where TMemberDeclarationSyntax : SyntaxNode
     {
-        internal const string DiagnosticId = "S4136";
-        protected const string MessageFormat = "All '{0}' method overloads should be adjacent.";
+        protected const string DiagnosticId = "S4136";
+        private const string MessageFormat = "All '{0}' method overloads should be adjacent.";
 
         private readonly DiagnosticDescriptor rule;
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
-
-        protected MethodOverloadsShouldBeGroupedBase(System.Resources.ResourceManager resources)
-        {
-            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, resources);
-        }
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+        protected abstract TSyntaxKind[] SyntaxKinds { get; }
 
+        protected abstract IEnumerable<TMemberDeclarationSyntax> GetMemberDeclarations(SyntaxNode node);
         protected abstract MemberInfo CreateMemberInfo(SyntaxNodeAnalysisContext c, TMemberDeclarationSyntax member);
 
-        protected void CheckMembers(SyntaxNodeAnalysisContext c, IEnumerable<TMemberDeclarationSyntax> members)
-        {
-            foreach (var misplacedOverload in GetMisplacedOverloads(c, members))
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected MethodOverloadsShouldBeGroupedBase() =>
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, Language.RspecResources);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
             {
-                var firstName = misplacedOverload.First().NameSyntax;
-                var secondaryLocations = misplacedOverload.Skip(1).Select(x => new SecondaryLocation(x.NameSyntax.GetLocation(), "Non-adjacent overload"));
-                c.ReportDiagnosticWhenActive(
-                    Diagnostic.Create(
-                        descriptor: rule,
-                        location: firstName.GetLocation(),
-                        additionalLocations: secondaryLocations.ToAdditionalLocations(),
-                        properties: secondaryLocations.ToProperties(),
-                        messageArgs: firstName.ValueText));
-            }
-        }
+                if (c.ContainingSymbol.Kind != SymbolKind.NamedType)
+                {
+                    return;
+                }
+
+                foreach (var misplacedOverload in GetMisplacedOverloads(c, GetMemberDeclarations(c.Node)))
+                {
+                    var firstName = misplacedOverload.First().NameSyntax;
+                    var secondaryLocations = misplacedOverload.Skip(1).Select(x => new SecondaryLocation(x.NameSyntax.GetLocation(), "Non-adjacent overload"));
```

Comment:
The `secondaryLocations` enumerable is iterated twice on report diagnostic call which means the `SecondaryLocation` classes are created twice. Could you please make it an array or a list?